### ISSUE TITLE
Percentiles and median

### ISSUE THREAD

**Issue by Mexx77** [graphite-ng/carbon-relay-ng#262]
Hey Dear carbon-relay-ng supporter,

I need percentiles and median aggregation functions. I know, its more complex and more memory intensive to compute these values (compared to mean). I need them anyway. Any chance to implement percentile aggregators? I guess one would do that in aggregator/processor.go, right?

**Comment by Dieterbe** [comment]
nowadays I think that code lives mostly in aggregator/processor.go

but yeah should be fairly easy to add.

**Comment by Mexx77** [comment]
I took a deep dive into the code that seemed relevant concerning a new aggregation function for percentiles. I guess that most people (including me) are usually interested in more than one percentile for a given metric. Say P50, P75, P90, P99, P999

If I follow the current design, I would need to configure 5 Aggregators, each firing up a Processor that holds all values of the interval and each sorting all values of the interval when being flushed. Quite a memory overhead when dealing with many data points per interval. In my case a problem.

It would increase performance significantly, if I had a single percentile Processor that returned a slice of structs, say a slice of processorResult:
`type processorResult struct {
    name string
    val  float64
}` with values [processorResult{name: "p99", val: 77}, processorResult{name: "p50", val: 25.3}]

In `func (a *Aggregator) Flush(ts uint)` I could than iterate over this slice and write each metric out to `metric := fmt.Sprintf("%s.%s %f %d", string(k.key), result.name, result.val, k.ts)`

As far as I am concerned this would have the following implications:
* I would have to adjust each Processor's Flush method to return a list of this struct (with most of them holding a single item)
* The current format string for aggregators would no longer be useful but
* Each Processor would hard code a suffix to the metric. E.g. avg would write "original.metric.name.avg"
* The user would no no longer be able to configure the resulting metric name via the format string

What do you think, is this an adjustment you would support and merge into the project? Any suggestions?

Thanks, Mexx

**Comment by Dieterbe** [comment]
you bring up a good point and i agree mostly with the proposition of having Processor.Flush return a slice.
I think the outFormat should stay. we just need to find an elegant way to make it work with percentiles.
we can make it work for example by letting users specify a format like `foo.bar.$1_$percentile`, where $1 gets filled in by the `match` function to determine the key (keeping $percentile literally), and expanding $percentile in the Aggregator.Flush step

**Comment by Mexx77** [comment]
Hey Dieterbe,
thank you for your feedback. It made me rethink my proposal and after a talk to my college, I would like to adjust my suggestion:

Using two different parameter types for outFormat (one for for regex group replacement and another one for the aggregation function names) makes the config cluttered and the code over complicated in my eyes. I think it should stay as it is with just the $1, $2 for regex group replacement. However, I would then let the Aggregator.Flush expand the slice as explained above and let it append a function name to the metric.

This would not break the API. A configuration like this one
```
[[aggregation]]
function = 'avg'
regex = '^stats\.timers\.(app|proxy|static)[0-9]+\.requests\.(.*)'
format = 'stats.timers._avg_$1.requests.$2'
```
would then result in metrics like `stats.timers._avg_app.requests.foo.avg`. A user could then, if preferred, adjust the format to `stats.timers.$1.requests.$2` giving him metrics like `stats.timers.app.requests.foo.avg`.

For a percentiles configuration:
```
[[aggregation]]
function = 'percentiles'
regex = '^stats\.timers\.(app|proxy|static)[0-9]+\.requests\.(.*)'
format = 'stats.timers.$1.requests.$2'
```
he would get metrics
* `stats.timers.app.requests.foo.p50`
* `stats.timers.app.requests.foo.p75`
* `stats.timers.app.requests.foo.p90`
* `stats.timers.app.requests.foo.p99`
* `stats.timers.app.requests.foo.p999`

In my opinion this is pretty clean. What do you think?

**Comment by Dieterbe** [comment]
so basically all regex/outFormat stuff stays exactly the same and the aggregator is just hardcoded to append those pxx values at the end of the string.
I think that is fine. In the future, some feature request might come along of someone wanting more control (in that case I think my proposal would be appropriate), but until then, I appreciate the simplicity of your proposal.

**Comment by Dieterbe** [comment]
I didn't examine your example closely enough. for the new pxx values I think it's ok to append them at the end, but for all the other pre-existing ones like avg, we can't just change the formatting and start adding '.avg' etc at the end. I'm pretty sure that's too rigid for some people, and also it would needlessly change the names of people's metrics which is too much trouble.

Instead, I will continue from your PR to keep the metrics consistent for the old aggregations

**Comment by Dieterbe** [comment]
merged via 5dd732ac2031ce232b7d7faf4d4d5172720387ac
